installinator/src/peers.rs

```diff
@@ -0,0 +1,334 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{fmt, future::Future, net::Ipv6Addr, str::FromStr, time::Duration};
+
+use anyhow::{bail, Result};
+use buf_list::BufList;
+use bytes::{Bytes, BytesMut};
+use display_error_chain::DisplayErrorChain;
+use itertools::Itertools;
+use progenitor_client::ResponseValue;
+use reqwest::StatusCode;
+use tokio::{
+    sync::{mpsc, oneshot},
+    time::Instant,
+};
+
+use crate::errors::{ArtifactFetchError, DiscoverPeersError};
+
+/// A fetched artifact
+pub(crate) struct FetchedArtifact {
+    pub(crate) attempt: usize,
+    pub(crate) addr: Ipv6Addr,
+    pub(crate) artifact: BufList,
+}
+
+impl FetchedArtifact {
+    /// In a loop, discover peers, and fetch from them.
+    ///
+    /// This only produces an error if the discover function errors out. In normal use, it is expected
+    /// to never error out.
+    pub(crate) async fn loop_fetch_from_peers<F, Fut>(
+        log: &slog::Logger,
+        mut discover_fn: F,
+        artifact_id: &ArtifactId,
+    ) -> Result<Self>
+    where
+        F: FnMut() -> Fut,
+        Fut: Future<Output = Result<Peers, DiscoverPeersError>>,
+    {
+        let mut attempt = 0;
+        loop {
+            attempt += 1;
+            let peers = match discover_fn().await {
+                Ok(peers) => peers,
+                Err(DiscoverPeersError::Retry(error)) => {
+                    slog::debug!(
+                        log,
+                        "(attempt {attempt}) failed to discover peers, retrying: {}",
+                        DisplayErrorChain::new(AsRef::<dyn std::error::Error>::as_ref(&error)),
+                    );
+                    // XXX: add a delay here?
```

Comment:
FWIW a fixed-duration async sleep on the order of milliseconds seems reasonable as a first pass, just to avoid slamming the CPU?

Reply:
Done -- added a 10ms delay.